### ISSUE TITLE
Clarinet unit tests revamp POC

### ISSUE THREAD

**Issue by hugocaillard** [hirosystems/clarinet#1052]
### Context

Some initial context in [this discussion](https://github.com/hirosystems/clarinet/discussions/1022)

TL;DR
- Clarinet currently embeds Deno to run `clarinet tests`
- This dependency is heavy, painful to maintain, and comes with build issues
- We decided to explore another approach for Clarinet test

### The POC

Clarinet will expose a JS library (through wasm) that will allow to test smart contracts in any JS/WASM environment (Node, Deno, Web Browers, etc).

- The package could be called `@stacks/clarinet-sdk/unit-tests` (introducing the clarinet SDK package #1034)
- The POC will be mostly tested with [Node 20 test runner](https://nodejs.org/api/test.html)

- here as a working template representing what's feasible at the end of this POC

```js
import { main } from "../../../../hiro/clarinet/components/clarinet-sdk/dist/index.js";
import { before, describe, it } from "node:test";
import assert from "node:assert/strict";
import { Cl } from "@stacks/transactions";

let sender = "ST1SJ3DTE5DN7X54YDH5D64R3BCB6A2AG2ZQ8YPD5";

describe("test counter", () => {
  let session;

  before(async () => {
    session = await main();
    await session.initSession(process.cwd(), "./Clarinet.toml");
  });

  it("gets counter value", () => {
    const res = session.callReadOnlyFn("counter", "get-counter", [], sender);
    assert.deepEqual(Cl.int(0), res.result);
    assert.equal(session.blockHeight, 1);
  });

  it("increments counter value", () => {
    const incrRes = session.callPublicFn("counter", "increment", [], sender);
    assert.deepEqual(incrRes.result, Cl.ok(Cl.bool(true)));
    assert.equal(session.blockHeight, 2);

    let res = session.callReadOnlyFn("counter", "get-counter", [], sender);
    assert.deepEqual(res.result, Cl.int(1));
    assert.equal(session.blockHeight, 2);
  });

  it("add any value", () => {
    const addRes = session.callPublicFn("counter", "add", [Cl.int(10)], sender);
    assert.deepEqual(addRes.result, Cl.ok(Cl.bool(true)));
    assert.equal(session.blockHeight, 3);

    let res = session.callReadOnlyFn("counter", "get-counter", [], sender);
    assert.deepEqual(res.result, Cl.int(11));
    assert.equal(session.blockHeight, 3);
  });
});
```

### Going further

A lot of features and improvements have to be considered in the next iteration (Q3). This is just a list of ideas, not all of them will be implemented
- specify which command will `clarinet test` run (default will be `node test`)
- throw errors on clarity run time error #998
- access to data-vars and maps #970
- access to private functions #547

**Comment by hugocaillard** [comment]
This POC is now considered done.
Conclusion:
- clarity sessions can easily be initiated in JS through wasm
- developer can easily interact with the session (call contracts, interact with raw clarity values and (de)serialize it with stacks.js)

Let's pursue in this direction to build a enjoyable testing experience!

Closing in favor of #1065
